### ISSUE TITLE
FormPost request using go 1.13 & HTTP/2 causes resteasy exception with TLS in 2.7.6.Final

### ISSUE THREAD

**Issue by DGuhr** [quarkusio/quarkus#26937]
### Describe the bug

When I do a `POST` request from go 1.13 on a quarkus 2.7.6 resteasy classic endpoint that consumes `MediaType.APPLICATION_FORM_URLENCODED` **and** uses SSL/TLS, the request causes an exception. See the added reproducer for more details.  

### Expected behavior

Response is sent normally, no exception occurs.

### Actual behavior

Exception: 
```
Caused by: java.lang.RuntimeException: java.io.IOException: Underlying input stream returned zero bytes
	at org.jboss.resteasy.plugins.server.BaseHttpRequest.getFormParameters(BaseHttpRequest.java:61)
	at org.jboss.resteasy.plugins.server.BaseHttpRequest.getDecodedFormParameters(BaseHttpRequest.java:74)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.ContextParameterInjector$GenericDelegatingProxy.invoke(ContextParameterInjector.java:166)
	at com.sun.proxy.$Proxy45.getDecodedFormParameters(Unknown Source)
	at com.example.ExampleResource.processPostFormRequest(ExampleResource.java:67)
	at com.example.ExampleResource_Subclass.processPostFormRequest$$superforward1(Unknown Source)
	at com.example.ExampleResource_Subclass$$function$$4.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	...
	
Caused by: java.io.IOException: Underlying input stream returned zero bytes
	at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:288)
	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.base/java.io.InputStreamReader.read(InputStreamReader.java:181)
	at java.base/java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.base/java.io.BufferedReader.read1(BufferedReader.java:212)
	at java.base/java.io.BufferedReader.read(BufferedReader.java:287)
	at org.jboss.resteasy.plugins.providers.FormUrlEncodedProvider.parseForm(FormUrlEncodedProvider.java:88)
	at org.jboss.resteasy.plugins.server.BaseHttpRequest.getFormParameters(BaseHttpRequest.java:57)
	... 48 more
```
, response not sent.

### How to Reproduce?

Reproducer: https://github.com/DGuhr/reproducer/tree/formPostRequest_ssl_exception

Steps: See Readme.md on that branch

### Output of `uname -a` or `ver`

Darwin dguhr-mac 20.6.0 Darwin Kernel Version 20.6.0: Tue Apr 19 21:04:45 PDT 2022; root:xnu-7195.141.29~1/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "11.0.15" 2022-04-19 OpenJDK Runtime Environment Temurin-11.0.15+10 (build 11.0.15+10) OpenJDK 64-Bit Server VM Temurin-11.0.15+10 (build 11.0.15+10, mixed mode)

### GraalVM version (if different from Java)

-

### Quarkus version or git rev

2.7.6.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)

### Additional information

The error is not happening with a newer go version like 1.16, but over at Keycloak we had some other issues not related to go but with the same exception, that I will list here, so maybe worth looking into it to make quarkus SSL / HTTP2 "more compatible":

- https://github.com/keycloak/keycloak/issues/11010 (windows related, could not reproduce here sadly)
- https://github.com/keycloak/keycloak/discussions/10931 (discussion of 11010)
- https://github.com/keycloak/keycloak/issues/13315 (Issue we created after experiencing the same and investigate it in the go context for our legacy operator)

Apart from that: Not sure if somehow related to #26878 but may be by chance that I found 2 HTTPS/SSL-related bugs in that short time frame 🥲

**Comment by DGuhr** [comment]
I don'T know why the quarkus-bot added env windows, but it is not. cc @geoand @gsmet  ;)

**Comment by gsmet** [comment]
I don't think this is related to Quarkus so maybe it makes more sense to open a RESTEasy issue? /cc @jamezp as he will know better.

**Comment by jamezp** [comment]
I'm not able to reproduce this, but I also don't have a Mac. I get the expected response:
```
Grant type is: password
```
The form parameters look correct too.

Go Environment:
```
go version go1.18.4 linux/amd64
```

Java Environment:
```
openjdk version "11.0.15" 2022-04-19
OpenJDK Runtime Environment 18.9 (build 11.0.15+10)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.15+10, mixed mode, sharing)
```

**Comment by vmuzikar** [comment]
@jamezp Could you please try building it with Go 1.13?

**Comment by jamezp** [comment]
I'll see if I can figure out how to do that. I'm not too familiar with Go :)

**Comment by DGuhr** [comment]
@jamezp Thanks for looking into it. I am also not sure if we could do sth about it. Just wanted to bring it to the attention of you because some users are also reporting the same exception to be thrown on windows (see referenced keycloak issues in additional information, not reproducible for me on my windows machine, though, so not entirely sure it is the same problem, but sounds hardly like it), so I thought maybe it is worth investigating a little further to see if we can do sth.

**Comment by jamezp** [comment]
Makes sense @DGuhr. There is potential to see if we really got any data and catch and maybe rethrow it. It seems like something should at least be logged as there really could be missing data. Given that we can file an issue at https://issues.redhat.com/browse/RESTEASY.

**Comment by DGuhr** [comment]
@jamezp sorry for reaching out back late, I had a few other things on my plate. Chiming through your board, I haven't found an issue yet, so created https://issues.redhat.com/browse/RESTEASY-3175 - feel free to modify/close if there is already another one.

A newer data point: Seems really not go-related but more general, see e.g. https://github.com/keycloak/keycloak/issues/13315#issuecomment-1203824676 (that's the Keycloak issue this issue derives from)

**Comment by pedroigor** [comment]
@jamezp After investigating more this issue, I think we were able to reproduce the issue locally using standalone Keycloak.

I'm not sure why it happens, but for some reason, this [handler](https://github.com/quarkusio/quarkus/blob/8f7492383d0de26345e690fb73906150805eb4d1/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxInputStream.java#L190) is only invoked after a call to `VertxInputStream.read`.

When the error occurs, the second call to `VertxInputStream.read` from [here](https://github.com/resteasy/Resteasy/blob/main/resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/FormUrlEncodedProvider.java#L88) is not yet notified about the end of the stream (`eof` == `false` because the handler was not yet called) hence return `0` and causing the exception.

I did not find yet why it works in certain cases and not for others. Perhaps this is related to the connection kept open by the client hence not invoking that handler that sets the `eof` to `true`?

To reproduce, please create a container image using this `Dockerfile`:

```
FROM quay.io/keycloak/keycloak:latest as builder

WORKDIR /opt/keycloak
# for demonstration purposes only, please make sure to use proper certificates in production instead
RUN keytool -genkeypair -storepass password -storetype PKCS12 -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore conf/server.keystore
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:latest
COPY --from=builder /opt/keycloak/ /opt/keycloak/

# change these values to point to a running postgres instance
ENV KC_HOSTNAME=localhost
ENV KC_CACHE=local
ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
```

To build the image:

```
docker build . -t keycloak-issue-26937
```

Then start the container as follows:

```
docker run --rm --name keycloak -p 8443:8443 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin keycloak-issue-26937 start
```

To reproduce the issue, you need to install [Burp Community](https://portswigger.net/burp) and use its browser:

* Open Burp
* Click on the `Proxy` tab
* Click the `Open browser` button

In the browser, you try to access Keycloak at `https://localhost:8443/admin`. You should be redirected to the login page.

On the login page, type `admin`/`admin` as the username and password.

The error should happen right after you click the button and you can see the stacktrace in the logs.

**Comment by jsorah** [comment]
I was able to inspect this further at the HTTP/2 protocol level using @DGuhr 's reproducer application and go client as well as using curl + burp proxy to the same reproducer application.  In both cases that fail it seems due to a second, empty DATA frame sent as the marker for the end stream.

In the scenarios that work over HTTP/2 - there is a single DATA frame with the end stream flag set containing the payload.

To perform this set up:
* cloned the reproducer repo - https://github.com/DGuhr/reproducer/tree/formPostRequest_ssl_exception
* Generated TLS certificates locally for the quarkus app (as per the reproducer readme)
* started the quarkus application `quarkus dev`
* Add extract-tls-secrets agent to extract the TLS secrets from the server for decryption - https://github.com/neykov/extract-tls-secrets
* Start wireshark locally, I targeted the loopback interface and set filter to `tcp.port == 8443`
  * Configure the SSL/TLS preferences to point at the log file from the extract-tls-secrets agent
* Compile the go client mentioned in the reproducer repo with go `1.13.15` 
* Execute the binary, and view results in wireshark
* Additionally, one can use curl + burp suite proxy to simulate this without the go client  
  * `curl --proxy 127.0.0.1:8081 -v --http2 -k -d "grant_type=http2test" https://localhost:8443/hello`

Unfortunately Wireshark is being difficult in exporting the decrypted packets I captured, so hoping these screenshots can help convey what I am observing.

Failing request:
![image](https://user-images.githubusercontent.com/4944041/188214052-ca71f9fa-4deb-4e82-b6b4-815ec84f6590.png)

Successful request:
![image](https://user-images.githubusercontent.com/4944041/188214250-6df9227d-f2c1-4275-86a3-5c2eb23f200d.png)

**Comment by pedroigor** [comment]
Is the second empty data frame somewhat expected when doing h2? Or is it a gap in `quarkus-vertx-http` or even in vert.x?

**Comment by jsorah** [comment]
> Is the second empty data frame somewhat expected when doing h2? Or is it a gap in `quarkus-vertx-http` or even in vert.x?
> 
> Also, do you have any clue why some clients do not send this second frame?

This is really my first deep dive into h2, so I'm not super confident in my answers here... but... I don't see anything in the spec that would disallow a DATA frame with empty payload and only flags. It may be inefficient (why not just add it on last frame w/ data instead of adding net new) - but seemingly legal.

I actually have yet _another_ example that can replicate it by explicitly creating the condition via - https://gist.github.com/jsorah/e80e55de3c340521518bf87d2a1710ce#file-empty_frame-py-L57

**Comment by vietj** [comment]
@jsorah can you provide a similar behavior with a vanilla Resteasy application (that uses Vert.x) and the VertxInputStream ? that would help a lot

**Comment by jamezp** [comment]
I'm not really sure this is a RESTEasy issue then. RESTEasy might be able to do some better reporting, but that's about it. Currently the default RESTEasy client does not support http/2.

**Comment by gsmet** [comment]
@cescoffier isn't this issue fixed by what you did here: https://github.com/quarkusio/quarkus/pull/27772 ?

**Comment by cescoffier** [comment]
@gsmet Yes. Sorry, forgot about that one (I mostly discussed on the keycloack issue).

**Comment by cescoffier** [comment]
Fixed by #22772.

**Comment by slominskir** [comment]
Just FYI - I'm seeing this Exception when a user attempts browser authentication using SPNEGO and their credential cache is empty (expired).   For example use kinit to repopulate the cache if empty (klist to see it) then try to login and  it works fine.   Use kdestroy and then try to reauthenticate and you'll get this Exception.
